File: auth0/src/main/java/com/auth0/android/provider/OAuthManager.java
package com.auth0.android.provider;

import android.app.Activity;
import android.net.Uri;
import android.text.TextUtils;
import android.util.Base64;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import com.auth0.android.Auth0;
import com.auth0.android.Auth0Exception;
import com.auth0.android.authentication.AuthenticationAPIClient;
import com.auth0.android.authentication.AuthenticationException;
import com.auth0.android.callback.BaseCallback;
import com.auth0.android.jwt.DecodeException;
import com.auth0.android.jwt.JWT;
import com.auth0.android.result.Credentials;

import java.security.SecureRandom;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

@SuppressWarnings("WeakerAccess")
class OAuthManager extends ResumableManager {

    private static final String TAG = OAuthManager.class.getSimpleName();

    static final String KEY_RESPONSE_TYPE = "response_type";
    static final String KEY_STATE = "state";
    static final String KEY_NONCE = "nonce";
    static final String KEY_MAX_AGE = "max_age";
    static final String KEY_CONNECTION = "connection";
    static final String RESPONSE_TYPE_CODE = "code";

    private static final String ERROR_VALUE_INVALID_CONFIGURATION = "a0.invalid_configuration";
    private static final String ERROR_VALUE_AUTHENTICATION_CANCELED = "a0.authentication_canceled";
    private static final String ERROR_VALUE_ACCESS_DENIED = "access_denied";
    private static final String ERROR_VALUE_UNAUTHORIZED = "unauthorized";
    private static final String ERROR_VALUE_LOGIN_REQUIRED = "login_required";
    private static final String ERROR_VALUE_ID_TOKEN_VALIDATION_FAILED = "Could not verify the ID token";
    private static final String METHOD_SHA_256 = "S256";
    private static final String KEY_CODE_CHALLENGE = "code_challenge";
    private static final String KEY_CODE_CHALLENGE_METHOD = "code_challenge_method";
    private static final String KEY_CLIENT_ID = "client_id";
    private static final String KEY_REDIRECT_URI = "redirect_uri";
    private static final String KEY_TELEMETRY = "auth0Client";
    private static final String KEY_ERROR = "error";
    private static final String KEY_ERROR_DESCRIPTION = "error_description";
    private static final String KEY_CODE = "code";

    private final Auth0 account;
    private final AuthCallback callback;
    private final Map<String, String> parameters;
    private final AuthenticationAPIClient apiClient;

    private int requestCode;
    private PKCE pkce;
    private Long currentTimeInMillis;
    private final CustomTabsOptions ctOptions;
    private Integer idTokenVerificationLeeway;
    private String idTokenVerificationIssuer;

    OAuthManager(@NonNull Auth0 account, @NonNull AuthCallback callback, @NonNull Map<String, String> parameters, @NonNull CustomTabsOptions ctOptions) {
        this.account = account;
        this.callback = callback;
        this.parameters = new HashMap<>(parameters);
        this.parameters.put(KEY_RESPONSE_TYPE, RESPONSE_TYPE_CODE);
        this.apiClient = new AuthenticationAPIClient(account);
        this.ctOptions = ctOptions;
    }

    @VisibleForTesting
    void setPKCE(@Nullable PKCE pkce) {
        this.pkce = pkce;
    }

    void setIdTokenVerificationLeeway(Integer leeway) {
        this.idTokenVerificationLeeway = leeway;
    }

    void setIdTokenVerificationIssuer(String issuer) {
        this.idTokenVerificationIssuer = TextUtils.isEmpty(issuer) ? apiClient.getBaseURL() : issuer;
    }

    void startAuthentication(Activity activity, String redirectUri, int requestCode) {
        addPKCEParameters(parameters, redirectUri);
        addClientParameters(parameters, redirectUri);
        addValidationParameters(parameters);
        Uri uri = buildAuthorizeUri();
        this.requestCode = requestCode;

        AuthenticationActivity.authenticateUsingBrowser(activity, uri, ctOptions);
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    boolean resume(AuthorizeResult result) {
        if (!result.isValid(requestCode)) {
            Log.w(TAG, "The Authorize Result is invalid.");
            return false;
        }

        if (result.isCanceled()) {
            //User cancelled the authentication
            AuthenticationException exception = new AuthenticationException(ERROR_VALUE_AUTHENTICATION_CANCELED, "The user closed the browser app and the authentication was canceled.");
            callback.onFailure(exception);
            return true;
        }

        final Map<String, String> values = CallbackHelper.getValuesFromUri(result.getIntentData());
        if (values.isEmpty()) {
            Log.w(TAG, "The response didn't contain any of these values: code, state");
            return false;
        }
        logDebug("The parsed CallbackURI contains the following values: " + values);

        try {
            assertNoError(values.get(KEY_ERROR), values.get(KEY_ERROR_DESCRIPTION));
            assertValidState(parameters.get(KEY_STATE), values.get(KEY_STATE));
        } catch (AuthenticationException e) {
            callback.onFailure(e);
            return true;
        }

        // response_type=code
        pkce.getToken(values.get(KEY_CODE), new SimpleAuthCallback(callback) {

            @Override
            public void onSuccess(@NonNull final Credentials credentials) {
                assertValidIdToken(credentials.getIdToken(), new VoidCallback() {
                    @Override
                    public void onSuccess(@Nullable Void ignored) {
                        callback.onSuccess(credentials);
                    }

                    @Override
                    public void onFailure(@Nullable Auth0Exception error) {
                        AuthenticationException wrappedError = new AuthenticationException(ERROR_VALUE_ID_TOKEN_VALIDATION_FAILED, error);
                        callback.onFailure(wrappedError);
                    }
                });
            }
        });
        return true;
    }

    private void assertValidIdToken(String idToken, final VoidCallback validationCallback) {
        if (TextUtils.isEmpty(idToken)) {
            validationCallback.onFailure(new TokenValidationException("ID token is required but missing"));
            return;
        }
        final JWT decodedIdToken;
        try {
            decodedIdToken = new JWT(idToken);
        } catch (DecodeException ignored) {
            validationCallback.onFailure(new TokenValidationException("ID token could not be decoded"));
            return;
        }

        BaseCallback<SignatureVerifier, TokenValidationException> signatureVerifierCallback = new BaseCallback<SignatureVerifier, TokenValidationException>() {

            @Override
            public void onFailure(@NonNull TokenValidationException error) {
                validationCallback.onFailure(error);
            }

            @Override
            public void onSuccess(@Nullable SignatureVerifier signatureVerifier) {
                //noinspection ConstantConditions
                IdTokenVerificationOptions options = new IdTokenVerificationOptions(idTokenVerificationIssuer, apiClient.getClientId(), signatureVerifier);
                String maxAge = parameters.get(KEY_MAX_AGE);
                if (!TextUtils.isEmpty(maxAge)) {
                    //noinspection ConstantConditions
                    options.setMaxAge(Integer.valueOf(maxAge));
                }
                options.setClockSkew(idTokenVerificationLeeway);
                options.setNonce(parameters.get(KEY_NONCE));
                options.setClock(new Date(getCurrentTimeInMillis()));
                try {
                    new IdTokenVerifier().verify(decodedIdToken, options);
                    logDebug("Authenticated using web flow");
                    validationCallback.onSuccess(null);
                } catch (TokenValidationException exc) {
                    validationCallback.onFailure(exc);
                }
            }
        };

        String tokenAlg = decodedIdToken.getHeader().get("alg");
        String tokenKeyId = decodedIdToken.getHeader().get("kid");
        SignatureVerifier.forAsymmetricAlgorithm(tokenKeyId, apiClient, signatureVerifierCallback);
    }

    private long getCurrentTimeInMillis() {
        return currentTimeInMillis != null ? currentTimeInMillis : System.currentTimeMillis();
    }

    @VisibleForTesting
    void setCurrentTimeInMillis(long currentTimeInMillis) {
        this.currentTimeInMillis = currentTimeInMillis;
    }

    //Helper Methods

    private void assertNoError(String errorValue, String errorDescription) throws AuthenticationException {
        if (errorValue == null) {
            return;
        }
        Log.e(TAG, "Error, access denied. Check that the required Permissions are granted and that the Application has this Connection configured in Auth0 Dashboard.");
        if (ERROR_VALUE_ACCESS_DENIED.equalsIgnoreCase(errorValue)) {
            throw new AuthenticationException(ERROR_VALUE_ACCESS_DENIED, "Permissions were not granted. Try again.");
        } else if (ERROR_VALUE_UNAUTHORIZED.equalsIgnoreCase(errorValue)) {
            throw new AuthenticationException(ERROR_VALUE_UNAUTHORIZED, errorDescription);
        } else if (ERROR_VALUE_LOGIN_REQUIRED.equals(errorValue)) {
            //Whitelist to allow SSO errors go through
            throw new AuthenticationException(errorValue, errorDescription);
        } else {
            throw new AuthenticationException(ERROR_VALUE_INVALID_CONFIGURATION, "The application isn't configured properly for the social connection. Please check your Auth0's application configuration");
        }
    }

    @VisibleForTesting
    static void assertValidState(@NonNull String requestState, @Nullable String responseState) throws AuthenticationException {
        if (!requestState.equals(responseState)) {
            Log.e(TAG, String.format("Received state doesn't match. Received %s but expected %s", responseState, requestState));
            throw new AuthenticationException(ERROR_VALUE_ACCESS_DENIED, "The received state is invalid. Try again.");
        }
    }

    private Uri buildAuthorizeUri() {
        Uri authorizeUri = Uri.parse(account.getAuthorizeUrl());
        Uri.Builder builder = authorizeUri.buildUpon();
        for (Map.Entry<String, String> entry : parameters.entrySet()) {
            builder.appendQueryParameter(entry.getKey(), entry.getValue());
        }
        Uri uri = builder.build();
        logDebug("Using the following Authorize URI: " + uri.toString());
        return uri;
    }

    private void addPKCEParameters(Map<String, String> parameters, String redirectUri) {
        createPKCE(redirectUri);
        String codeChallenge = pkce.getCodeChallenge();
        parameters.put(KEY_CODE_CHALLENGE, codeChallenge);
        parameters.put(KEY_CODE_CHALLENGE_METHOD, METHOD_SHA_256);
        Log.v(TAG, "Using PKCE authentication flow");
    }

    private void addValidationParameters(Map<String, String> parameters) {
        String state = getRandomString(parameters.get(KEY_STATE));
        String nonce = getRandomString(parameters.get(KEY_NONCE));
        parameters.put(KEY_STATE, state);
        parameters.put(KEY_NONCE, nonce);
    }

    private void addClientParameters(Map<String, String> parameters, String redirectUri) {
        if (account.getTelemetry() != null) {
            parameters.put(KEY_TELEMETRY, account.getTelemetry().getValue());
        }
        parameters.put(KEY_CLIENT_ID, account.getClientId());
        parameters.put(KEY_REDIRECT_URI, redirectUri);
    }

    private void createPKCE(String redirectUri) {
        if (pkce == null) {
            pkce = new PKCE(apiClient, redirectUri);
        }
    }

    @VisibleForTesting
    static String getRandomString(@Nullable String defaultValue) {
        return defaultValue != null ? defaultValue : secureRandomString();
    }

    private static String secureRandomString() {
        final SecureRandom sr = new SecureRandom();
        final byte[] randomBytes = new byte[32];
        sr.nextBytes(randomBytes);
        return Base64.encodeToString(randomBytes, Base64.URL_SAFE | Base64.NO_WRAP | Base64.NO_PADDING);
    }

    private void logDebug(String message) {
        if (account.isLoggingEnabled()) {
            Log.d(TAG, message);
        }
    }
}

File: auth0/src/main/java/com/auth0/android/provider/OAuthManager.kt
package com.auth0.android.provider

import android.content.Context
import android.net.Uri
import android.text.TextUtils
import android.util.Base64
import android.util.Log
import androidx.annotation.VisibleForTesting
import com.auth0.android.Auth0
import com.auth0.android.Auth0Exception
import com.auth0.android.authentication.AuthenticationAPIClient
import com.auth0.android.authentication.AuthenticationException
import com.auth0.android.callback.Callback
import com.auth0.android.dpop.DPoPProvider
import com.auth0.android.dpop.KeyStoreManager
import com.auth0.android.request.internal.Jwt
import com.auth0.android.request.internal.OidcUtils
import com.auth0.android.result.Credentials
import java.security.SecureRandom
import java.util.*

internal class OAuthManager(
    private val account: Auth0,
    private val callback: Callback<Credentials, AuthenticationException>,
    parameters: Map<String, String>,
    ctOptions: CustomTabsOptions,
    private val launchAsTwa: Boolean = false,
    private val customAuthorizeUrl: String? = null
) : ResumableManager() {
    private val parameters: MutableMap<String, String>
    private val headers: MutableMap<String, String>
    private val ctOptions: CustomTabsOptions
    private val apiClient: AuthenticationAPIClient
    private var requestCode = 0
    private var pkce: PKCE? = null

    private val dPoPProvider = DPoPProvider()
    private val keyStoreManager = KeyStoreManager()

    init {
        if (keyStoreManager.getEs256KeyPair() == null) {
            keyStoreManager.generateKeyPair()
        }
    }

    private var _currentTimeInMillis: Long? = null

    @set:VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    internal var currentTimeInMillis: Long
        get() = if (_currentTimeInMillis != null) _currentTimeInMillis!! else System.currentTimeMillis()
        set(value) {
            _currentTimeInMillis = value
        }

    private var idTokenVerificationLeeway: Int? = null
    private var idTokenVerificationIssuer: String? = null

    @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
    fun setPKCE(pkce: PKCE?) {
        this.pkce = pkce
    }

    fun setIdTokenVerificationLeeway(leeway: Int?) {
        idTokenVerificationLeeway = leeway
    }

    fun setIdTokenVerificationIssuer(issuer: String?) {
        idTokenVerificationIssuer = if (TextUtils.isEmpty(issuer)) apiClient.baseURL else issuer
    }

    fun startAuthentication(context: Context, redirectUri: String, requestCode: Int) {
        OidcUtils.includeDefaultScope(parameters)
        addPKCEParameters(parameters, redirectUri, headers)
        addClientParameters(parameters, redirectUri)
        addValidationParameters(parameters)
        addDopParameter(parameters)
        val uri = buildAuthorizeUri()
        this.requestCode = requestCode
        AuthenticationActivity.authenticateUsingBrowser(context, uri, launchAsTwa, ctOptions)
    }

    fun setHeaders(headers: Map<String, String>) {
        this.headers.putAll(headers)
        dPoPProvider.generateDpopProofJwt("POST", "https://${account.domain}/oauth/token")?.let {
            this.headers["DPoP"] = it
        }
    }

    public override fun resume(result: AuthorizeResult): Boolean {
        if (!result.isValid(requestCode)) {
            Log.w(TAG, "The Authorize Result is invalid.")
            return false
        }
        if (result.isCanceled) {
            //User cancelled the authentication
            val exception = AuthenticationException(
                AuthenticationException.ERROR_VALUE_AUTHENTICATION_CANCELED,
                "The user closed the browser app and the authentication was canceled."
            )
            callback.onFailure(exception)
            return true
        }
        val values = CallbackHelper.getValuesFromUri(result.intentData)
        if (values.isEmpty()) {
            Log.w(TAG, "The response didn't contain any of these values: code, state")
            return false
        }
        Log.d(TAG, "The parsed CallbackURI contains the following parameters: ${values.keys}")
        try {
            assertNoError(values[KEY_ERROR], values[KEY_ERROR_DESCRIPTION])
            assertValidState(parameters[KEY_STATE]!!, values[KEY_STATE])
        } catch (e: AuthenticationException) {
            callback.onFailure(e)
            return true
        }

        // response_type=code
        pkce!!.getToken(
            values[KEY_CODE],
            object : Callback<Credentials, AuthenticationException> {
                override fun onSuccess(credentials: Credentials) {
                    assertValidIdToken(
                        credentials.idToken,
                        object : Callback<Void?, Auth0Exception> {
                            override fun onSuccess(result: Void?) {
                                callback.onSuccess(credentials)
                            }

                            override fun onFailure(error: Auth0Exception) {
                                val wrappedError = AuthenticationException(
                                    ERROR_VALUE_ID_TOKEN_VALIDATION_FAILED, error
                                )
                                callback.onFailure(wrappedError)
                            }
                        })
                }

                override fun onFailure(error: AuthenticationException) {
                    if ("Unauthorized" == error.getDescription()) {
                        Log.e(
                            PKCE.TAG,
                            "Unable to complete authentication with PKCE. PKCE support can be enabled by setting Application Type to 'Native' and Token Endpoint Authentication Method to 'None' for this app at 'https://manage.auth0.com/#/applications/" + apiClient.clientId + "/settings'."
                        )
                    }
                    callback.onFailure(error)
                }
            })
        return true
    }

    public override fun failure(exception: AuthenticationException) {
        callback.onFailure(exception)
    }

    private fun assertValidIdToken(
        idToken: String?,
        validationCallback: Callback<Void?, Auth0Exception>
    ) {
        if (TextUtils.isEmpty(idToken)) {
            validationCallback.onFailure(IdTokenMissingException())
            return
        }
        val decodedIdToken: Jwt = try {
            Jwt(idToken!!)
        } catch (error: Exception) {
            validationCallback.onFailure(
                UnexpectedIdTokenException(error)
            )
            return
        }
        val signatureVerifierCallback: Callback<SignatureVerifier, TokenValidationException> =
            object : Callback<SignatureVerifier, TokenValidationException> {
                override fun onFailure(error: TokenValidationException) {
                    validationCallback.onFailure(error)
                }

                override fun onSuccess(result: SignatureVerifier) {
                    val options = IdTokenVerificationOptions(
                        idTokenVerificationIssuer!!,
                        apiClient.clientId,
                        result
                    )
                    val maxAge = parameters[KEY_MAX_AGE]
                    if (!TextUtils.isEmpty(maxAge)) {
                        options.maxAge = Integer.valueOf(maxAge!!)
                    }

                    options.clockSkew = idTokenVerificationLeeway
                    options.nonce = parameters[KEY_NONCE]
                    options.clock = Date(currentTimeInMillis)
                    options.organization = parameters[KEY_ORGANIZATION]
                    try {
                        IdTokenVerifier().verify(decodedIdToken, options, true)
                        validationCallback.onSuccess(null)
                    } catch (exc: TokenValidationException) {
                        validationCallback.onFailure(exc)
                    }
                }
            }
        val tokenKeyId = decodedIdToken.keyId
        SignatureVerifier.forAsymmetricAlgorithm(tokenKeyId, apiClient, signatureVerifierCallback)
    }

    internal fun toState(): OAuthManagerState {
        return OAuthManagerState(
            parameters = parameters.toMap(),
            headers = headers.toMap(),
            requestCode = requestCode,
            ctOptions = ctOptions,
            pkce = pkce,
            auth0 = account,
            idTokenVerificationIssuer = idTokenVerificationIssuer,
            idTokenVerificationLeeway = idTokenVerificationLeeway,
            customAuthorizeUrl = this.customAuthorizeUrl
        )
    }

    //Helper Methods
    @Throws(AuthenticationException::class)
    private fun assertNoError(errorValue: String?, errorDescription: String?) {
        if (errorValue == null) {
            return
        }
        Log.e(
            TAG,
            "Error, access denied. Check that the required Permissions are granted and that the Application has this Connection configured in Auth0 Dashboard."
        )
        val unknownErrorDescription = "An unexpected error occurred."
        when {
            ERROR_VALUE_ACCESS_DENIED.equals(errorValue, ignoreCase = true) -> {
                throw AuthenticationException(
                    ERROR_VALUE_ACCESS_DENIED,
                    errorDescription ?: "Permissions were not granted. Try again."
                )
            }

            ERROR_VALUE_UNAUTHORIZED.equals(errorValue, ignoreCase = true) -> {
                throw AuthenticationException(
                    ERROR_VALUE_UNAUTHORIZED,
                    errorDescription ?: unknownErrorDescription
                )
            }

            ERROR_VALUE_LOGIN_REQUIRED == errorValue -> {
                //Whitelist to allow SSO errors go through
                throw AuthenticationException(
                    errorValue,
                    errorDescription ?: unknownErrorDescription
                )
            }

            else -> {
                throw AuthenticationException(
                    errorValue,
                    errorDescription ?: unknownErrorDescription
                )
            }
        }
    }

    private fun buildAuthorizeUri(): Uri {
        val urlToUse = customAuthorizeUrl ?: account.authorizeUrl
        val authorizeUri = Uri.parse(urlToUse)
        val builder = authorizeUri.buildUpon()
        for ((key, value) in parameters) {
            builder.appendQueryParameter(key, value)
        }
        val uri = builder.build()
        Log.d(TAG, "Using the following Authorize URI: $uri")
        return uri
    }

    private fun addPKCEParameters(
        parameters: MutableMap<String, String>,
        redirectUri: String,
        headers: Map<String, String>
    ) {
        createPKCE(redirectUri, headers)
        val codeChallenge = pkce!!.codeChallenge
        parameters[KEY_CODE_CHALLENGE] = codeChallenge
        parameters[KEY_CODE_CHALLENGE_METHOD] = METHOD_SHA_256
        Log.v(TAG, "Using PKCE authentication flow")
    }

    private fun addValidationParameters(parameters: MutableMap<String, String>) {
        val state = getRandomString(parameters[KEY_STATE])
        val nonce = getRandomString(parameters[KEY_NONCE])
        parameters[KEY_STATE] = state
        parameters[KEY_NONCE] = nonce
    }

    private fun addDopParameter(parameters: MutableMap<String, String>) {
        dPoPProvider.getDpopJktValue()?.let {
            parameters["dpop_jkt"] = it
        }
    }

    private fun addClientParameters(parameters: MutableMap<String, String>, redirectUri: String) {
        parameters[KEY_AUTH0_CLIENT_INFO] = account.auth0UserAgent.value
        parameters[KEY_CLIENT_ID] = account.clientId
        parameters[KEY_REDIRECT_URI] = redirectUri
    }

    private fun createPKCE(redirectUri: String, headers: Map<String, String>) {
        if (pkce == null) {
            pkce = PKCE(apiClient, redirectUri, headers)
        }
    }

    companion object {
        private val TAG = OAuthManager::class.java.simpleName
        const val KEY_RESPONSE_TYPE = "response_type"
        const val KEY_STATE = "state"
        const val KEY_NONCE = "nonce"
        const val KEY_MAX_AGE = "max_age"
        const val KEY_CONNECTION = "connection"
        const val KEY_ORGANIZATION = "organization"
        const val KEY_INVITATION = "invitation"
        const val KEY_SCOPE = "scope"
        const val RESPONSE_TYPE_CODE = "code"
        private const val DEFAULT_SCOPE = "openid profile email"
        private const val REQUIRED_SCOPE = "openid"
        private const val ERROR_VALUE_INVALID_CONFIGURATION = "a0.invalid_configuration"
        private const val ERROR_VALUE_ACCESS_DENIED = "access_denied"
        private const val ERROR_VALUE_UNAUTHORIZED = "unauthorized"
        private const val ERROR_VALUE_LOGIN_REQUIRED = "login_required"
        private const val ERROR_VALUE_ID_TOKEN_VALIDATION_FAILED = "Could not verify the ID token"
        private const val METHOD_SHA_256 = "S256"
        private const val KEY_CODE_CHALLENGE = "code_challenge"
        private const val KEY_CODE_CHALLENGE_METHOD = "code_challenge_method"
        private const val KEY_CLIENT_ID = "client_id"
        private const val KEY_REDIRECT_URI = "redirect_uri"
        private const val KEY_AUTH0_CLIENT_INFO = "auth0Client"
        private const val KEY_ERROR = "error"
        private const val KEY_ERROR_DESCRIPTION = "error_description"
        private const val KEY_CODE = "code"

        @JvmStatic
        @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
        @Throws(AuthenticationException::class)
        fun assertValidState(requestState: String, responseState: String?) {
            if (requestState != responseState) {
                Log.e(
                    TAG,
                    String.format(
                        "Received state doesn't match. Received %s but expected %s",
                        responseState,
                        requestState
                    )
                )
                throw AuthenticationException(
                    ERROR_VALUE_ACCESS_DENIED,
                    "The received state is invalid. Try again."
                )
            }
        }

        @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
        fun getRandomString(defaultValue: String?): String {
            return defaultValue ?: secureRandomString()
        }

        private fun secureRandomString(): String {
            val sr = SecureRandom()
            val randomBytes = ByteArray(32)
            sr.nextBytes(randomBytes)
            return Base64.encodeToString(
                randomBytes,
                Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING
            )
        }
    }

    init {
        headers = HashMap()
        this.parameters = parameters.toMutableMap()
        this.parameters[KEY_RESPONSE_TYPE] = RESPONSE_TYPE_CODE
        apiClient = AuthenticationAPIClient(account)
        this.ctOptions = ctOptions
    }
}

internal fun OAuthManager.Companion.fromState(
    state: OAuthManagerState,
    callback: Callback<Credentials, AuthenticationException>
): OAuthManager {
    return OAuthManager(
        account = state.auth0,
        ctOptions = state.ctOptions,
        parameters = state.parameters,
        callback = callback,
        customAuthorizeUrl = state.customAuthorizeUrl
    ).apply {
        setHeaders(
            state.headers
        )
        setPKCE(state.pkce)
        setIdTokenVerificationIssuer(state.idTokenVerificationIssuer)
        setIdTokenVerificationLeeway(state.idTokenVerificationLeeway)
    }
}